controllers/storageclaim_controller.go

```diff
@@ -80,54 +111,74 @@ func (r *StorageClassClaimReconciler) SetupWithManager(mgr ctrl.Manager) error {
+		For(&v1alpha1.StorageClaim{}, builder.WithPredicates(
 			predicate.GenerationChangedPredicate{},
 		)).
 		Watches(&storagev1.StorageClass{}, enqueueStorageConsumerRequest).
 		Watches(&snapapi.VolumeSnapshotClass{}, enqueueStorageConsumerRequest).
+		Watches(&v1alpha1.StorageClassClaim{}, enqueueStorageClassClaimRequest, builder.WithPredicates(predicate.GenerationChangedPredicate{})).
 		Complete(r)
 }
 
+//+kubebuilder:rbac:groups=ocs.openshift.io,resources=storageclaims,verbs=get;list;watch;create;update;patch;delete
+//+kubebuilder:rbac:groups=ocs.openshift.io,resources=storageclaims/status,verbs=get;update;patch
+//+kubebuilder:rbac:groups=ocs.openshift.io,resources=storageclaims/finalizers,verbs=update
 //+kubebuilder:rbac:groups=ocs.openshift.io,resources=storageclassclaims,verbs=get;list;watch;create;update;patch;delete
 //+kubebuilder:rbac:groups=ocs.openshift.io,resources=storageclassclaims/status,verbs=get;update;patch
 //+kubebuilder:rbac:groups=ocs.openshift.io,resources=storageclassclaims/finalizers,verbs=update
 //+kubebuilder:rbac:groups="",resources=secrets,verbs=get;list;watch;create;update;delete
 //+kubebuilder:rbac:groups=storage.k8s.io,resources=storageclasses,verbs=get;list;watch;create;update;patch;delete
 //+kubebuilder:rbac:groups=snapshot.storage.k8s.io,resources=volumesnapshotclasses,verbs=get;list;watch;create;delete
 //+kubebuilder:rbac:groups=core,resources=persistentvolumes,verbs=get;list;watch
+//+kubebuilder:rbac:groups=snapshot.storage.k8s.io,resources=volumesnapshotcontents,verbs=get;list;watch
 
 // Reconcile is part of the main kubernetes reconciliation loop which aims to
 // move the current state of the cluster closer to the desired state.
 // TODO(user): Modify the Reconcile function to compare the state specified by
-// the StorageClassClaim object against the actual cluster state, and then
+// the StorageClaim object against the actual cluster state, and then
 // perform operations to make the cluster state reflect the state specified by
 // the user.
 //
 // For more details, check Reconcile and its Result here:
 // - https://pkg.go.dev/sigs.k8s.io/controller-runtime@v0.12.2/pkg/reconcile
-func (r *StorageClassClaimReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+func (r *StorageClaimReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
 
-	r.log = ctrllog.FromContext(ctx, "StorageClassClaim", req)
+	r.log = ctrllog.FromContext(ctx, "StorageClaim", req)
 	r.ctx = ctrllog.IntoContext(ctx, r.log)
-	r.log.Info("Reconciling StorageClassClaim.")
+	r.log.Info("Reconciling StorageClaim.")
+
+	// perform a migration of all existing storageclassclaims to storageclaims and have reconciler bring the
+	// resources for newly created storageclaims to the desired state
+	if err := r.migrateStorageClassClaims(); err != nil {
+		return reconcile.Result{}, fmt.Errorf("failed to migrate storageclassclaims")
+	}
```

Comment:
Are you planning to run the migration of all storageclassclaims for every request on a single claim? That seems like a problematic approach. Preferably we need a preprocess that runs before the controller starts (or as part of main) that does the migration only once.

Reply:
> migration of all storageclassclaims
- the intention is to migrate all and delete, so that from next cache update the process is mostly no op.

> does the migration only once
- what if the existing storageclassclaims migration fails due to some transient error and so this is part of reconcile and after migration is successful it'll be close to a no-op until next release.

Reply:
no, it should rerun

Reply:
So not changing.

Reply:
I think you are missing the point. A reconcile even for a storageclaim is for a single claim, not all claims. This code run for all of the old storageclassclaims which suggest that it should not be here but at a scope that only runs once

Reply:
moved to a mini controller as discussed. done.